sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala

```diff
@@ -97,8 +97,15 @@ class CatalogImpl(sparkSession: SparkSession) extends Catalog {
    */
   @throws[AnalysisException]("database does not exist")
   override def listTables(dbName: String): Dataset[Table] = {
-    val tables = sessionCatalog.listTables(dbName).map(makeTable)
-    CatalogImpl.makeDataset(tables, sparkSession)
+    if (sessionCatalog.databaseExists(dbName)) {
+      val tables = sessionCatalog.listTables(dbName).map(makeTable)
+      CatalogImpl.makeDataset(tables, sparkSession)
+    } else {
+      val multiParts = sparkSession.sessionState.sqlParser.parseMultipartIdentifier(dbName)
+      val plan = ShowTables(UnresolvedNamespace(multiParts), None)
+      val ret = sparkSession.sessionState.executePlan(plan).toRdd.collect()
+      CatalogImpl.makeDataset(Seq(), sparkSession)
```

Comment:
@cloud-fan 

how to convert `InternalRow` to `Table`?

Reply:
Are you saying we don't construct a logical plan for `listTables`? 

If so, now I am confused on when we should construct a logical plan (thus to match SQL command) and when we don't....